### ISSUE TITLE
"Failed to get language for textDocument/diagnostic" when saving a new file

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#81410]
Repro steps in VS Code:
1) New Untitled Text File (ctrl+N).
2) Change Language Mode to C#. No need to type anything.
3) Save as `MyFile.cs` or anything you want.

Expected behavior: Editing session proceeds without issues.
Actual behavior: An error is reported, and completions etc. don't work until running Restart Extension Host. Log:
```log
2025-11-21 19:22:01.982 [error] [textDocument/diagnostic] [LSP] Failed to get language for untitled:Untitled-1 with language 
2025-11-21 19:22:01.998 [error] [textDocument/diagnostic] [LSP] System.Exception: Failed to get language for textDocument/diagnostic
2025-11-21 19:22:02.003 [error] Request textDocument/diagnostic failed.
  Message: Failed to get language for textDocument/diagnostic
  Code: -32000 
[object Object]
2025-11-21 19:22:02.004 [error] Document pull failed for text document untitled:Untitled-1
  Message: Failed to get language for textDocument/diagnostic
  Code: -32000 
[object Object]
```

The content of the failing request is `{"identifier":"HotReloadDiagnostics","textDocument":{"uri":"untitled:Untitled-1"}}`

This bug occurs regardless of the `dotnet.projects.enableFileBasedPrograms` setting.
This bug occurs in the client regardless of whether CDK is enabled or disabled.

**Comment by RikkiGibson** [comment]
@dibarbet pretty sure you have seen this before, but, here is a straightforward repro. Maybe we can come up with an automated test which repros this.

**Comment by DrEmpiricism** [comment]
Having this same issue when I copy C# code from another editor over to VS and then save it.

**Comment by dibarbet** [comment]
This happens because of the following operations on save:
1.  VSCode sends a didClose for the 'untitled' document, followed by a didOpen for the new file URI.
2.  The Roslyn server handles the didClose by clearing out the saved state for the 'untitled' document, including the language.
3.  VSCode may send a textDocument/diagnostics request for the untitled document *after* it the didClose (most likely to clear diagnostics).  When the server gets the request, it cannot handle it because the language information associated with the 'untitled' URI no longer exists.

The only way I can think of to resolve this is to relax the language check, and default to 'C#' for any document that is not open.

**Comment by RikkiGibson** [comment]
> VSCode may send a textDocument/diagnostics request for the untitled document after it the didClose (most likely to clear diagnostics).

Is this allowed by the LSP standard?

edit: It looks like yes.
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didClose
> Note that a server’s ability to fulfill requests is independent of whether a text document is open or closed.
